x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentNodeService.java

```diff
@@ -188,39 +197,68 @@ void loadQueuedModels() {
             );
             if (unassignedIndices.size() > 0) {
                 logger.trace("not loading models as indices {} primary shards are unassigned", unassignedIndices);
+                onFinish.run();
                 return;
             }
         }
-        logger.trace("attempting to load all currently queued models");
-        // NOTE: As soon as this method exits, the timer for the scheduler starts ticking
-        Deque<TrainedModelDeploymentTask> loadingToRetry = new ArrayDeque<>();
-        while ((loadingTask = loadingModels.poll()) != null) {
-            final String deploymentId = loadingTask.getDeploymentId();
-            if (loadingTask.isStopped()) {
-                if (logger.isTraceEnabled()) {
-                    String reason = loadingTask.stoppedReason().orElse("_unknown_");
-                    logger.trace("[{}] attempted to load stopped task with reason [{}]", deploymentId, reason);
+
+        // deploy the models synchronously by flattening loadingModels into a chain of listeners
+        // if we need to retry, wait to add it back to loadingModels until after we finish this batch of deployments
+        // if someone calls stop halfway through, abandon this entire chain
+        var loadingToRetry = new ConcurrentLinkedDeque<TrainedModelDeploymentTask>();
+        var deploymentChain = SubscribableListener.<Void>newSucceeded(null);
+        while (loadingModels.isEmpty() == false) {
```

Comment:
Would it simplify the code to remove the `while()` loop in favour of just taking the head of the queue and loading the next model on the next scheduled invocation?

```
var loadingTask = loadingModels.poll();
if (loadingTask == null) {
  onFinish.run();
  return;
}

```

This function `loadQueuedModels()` will be called again periodically anyway so it's an option to let next invocation handle the next model. Would an error loading 1 model fail any other models in this chain?

In practice there are only a small number of models per ml node due to the memory and CPU demands. In an autoscaling situation when an ml node joins the cluster it may have to load 1 or 2 models, certainly not 10s of models.